webapp/components/compat-2021.js

```diff
@@ -6,10 +6,152 @@
+    return stable ?
+      this.stableDatatables.get(feature) :
+      this.experimentalDatatables.get(feature);
+  }
+
+  // Fetches a list of browser versions for stable or experimental. This is a
+  // helper method for building tooltip actions; the returned list has one
+  // entry per row in the corresponding datatables.
+  async getBrowserVersions(stable) {
+    await this._dataLoaded;
+    return stable ?
+      this.stableBrowserVersions :
+      this.experimentalBrowserVersions;
+  }
+
+  // Loads the unified CSV file for either stable or experimental, and
+  // processes it into the set of datatables provided by this class. Will
+  // ultimately set either this.stableDatatables or this.experimentalDatatables
+  // with a map of {feature name --> datatable}.
+  async _loadCsv(stable) {
+    const label = stable ? 'stable' : 'experimental';
+    const url = `${GITHUB_URL_PREFIX}/data/compat2021/unified-scores-${label}.csv`;
+    const csvLines = await fetchCsvContents(url);
+
+    const features = [SUMMARY_FEATURE_NAME, ...FEATURES];
+    const datatables = new Map(features.map(feature => {
+      const dataTable = new window.google.visualization.DataTable();
+      dataTable.addColumn('date', 'Date');
+      dataTable.addColumn('number', 'Chrome/Edge');
+      dataTable.addColumn({type: 'string', role: 'tooltip'});
+      dataTable.addColumn('number', 'Firefox');
+      dataTable.addColumn({type: 'string', role: 'tooltip'});
+      dataTable.addColumn('number', 'Safari');
+      dataTable.addColumn({type: 'string', role: 'tooltip'});
+      return [feature, dataTable];
+    }));
+
+    // We list Chrome/Edge on the legend, but when creating the tooltip we
+    // include the version information and so should be clear about which browser
+    // exactly gave the results.
+    const tooltipBrowserNames = [
+      'Chrome',
+      'Firefox',
+      'Safari',
+    ];
+
+    // We store a lookup table of browser versions to help with the 'show
+    // revision changelog' tooltip action.
+    const browserVersions = [[], [], []];
+
+    csvLines.forEach(line => {
+      // We control the CSV data source, so are quite lazy with parsing it.
+      //
+      // The format is:
+      //   date, [browser-version, browser-feature-a, browser-feature-b, ...]+
+      const csvValues = line.split(',');
+
+      // Javascript Date objects use 0-indexed months whilst the CSV is
+      // 1-indexed, so adjust for that.
+      const dateParts = csvValues[0].split('-').map(x => parseInt(x));
+      const date = new Date(dateParts[0], dateParts[1] - 1, dateParts[2]);
+
+      // Initialize a new row for each feature, with the date column set.
+      const newRows = new Map(features.map(feature => {
+        return [feature, [date]];
+      }));
+
+      // Now handle each of the browsers. For each there is a version column,
+      // then the scores for each of the five features.
+      for (let i = 1; i < csvValues.length; i += 6) {
```

Comment:
This loop works, but if there's a way to instead loop over `browserVersions` the right number of times it could end up more readable. I wouldn't know without trying it, though, so feel free to ignore if you don't think so. Not important.

Reply:
Yeah, there are better ways to do this but I think this is ok for now (no time!).